File: zeroinstall/cmd/import.py
"""
The B{0install import} command-line interface.
"""

# Copyright (C) 2011, Thomas Leonard
# See the README file for details, or visit http://0install.net.

import os

from zeroinstall import SafeException, _, logger
from zeroinstall.cmd import UsageError
from zeroinstall.injector import gpg
from zeroinstall.injector.iface_cache import PendingFeed
from zeroinstall.support import tasks
from xml.dom import minidom

syntax = "FEED"

def add_options(parser):
	pass

def handle(config, options, args):
	if not args:
		raise UsageError()

	for x in args:
		if not os.path.isfile(x):
			raise SafeException(_("File '%s' does not exist") % x)
		logger.info(_("Importing from file '%s'"), x)
		with open(x, 'rb') as signed_data:
			data, sigs = gpg.check_stream(signed_data)
			doc = minidom.parseString(data.read())
			uri = doc.documentElement.getAttribute('uri')
			if not uri:
				raise SafeException(_("Missing 'uri' attribute on root element in '%s'") % x)
			logger.info(_("Importing information about interface %s"), uri)
			signed_data.seek(0)

			pending = PendingFeed(uri, signed_data)

			def run():
				keys_downloaded = tasks.Task(pending.download_keys(config.fetcher), "download keys")
				yield keys_downloaded.finished
				tasks.check(keys_downloaded.finished)
				if not config.iface_cache.update_feed_if_trusted(uri, pending.sigs, pending.new_xml):
					blocker = config.trust_mgr.confirm_keys(pending)
					if blocker:
						yield blocker
						tasks.check(blocker)
					if not config.iface_cache.update_feed_if_trusted(uri, pending.sigs, pending.new_xml):
						raise SafeException(_("No signing keys trusted; not importing"))

			task = tasks.Task(run(), "import feed")

			tasks.wait_for_blocker(task.finished)

def complete(completion, args, cword):
	"""@type completion: L{zeroinstall.cmd._Completion}
	@type args: [str]
	@type cword: int"""
	if len(args) != 1: return
	completion.expand_files()

File: ocaml/import.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

(** The "0install import" command *)

open Options
open Zeroinstall.General
open Support
open Support.Common

module Q = Support.Qdom
module FeedAttr = Zeroinstall.Constants.FeedAttr
module U = Support.Utils

let import_feed options arg =
  let tools = options.tools in
  let system = options.config.system in

  if not (system#file_exists arg) then
    Safe_exn.failf "File '%s' does not exist" arg;

  log_info "Importing from file '%s'" arg;
  let xml = U.read_file system arg in
  let root = `String (0, xml) |> Xmlm.make_input |> Q.parse_input (Some arg) in
  let url = ZI.get_attribute FeedAttr.uri root in
  let parsed_url = match Zeroinstall.Feed_url.parse_non_distro url with
    | `Remote_feed _ as url -> url
    | `Local_feed _ -> Safe_exn.failf "Invalid URI '%s' on feed" url in

  log_info "Importing feed %s" url;

  Lwt_main.run ((tools#make_fetcher tools#ui#watcher)#import_feed parsed_url xml)

let handle options flags args =
  Support.Argparse.iter_options flags (function
    | #common_option as o -> Common_options.process_common_option options o
  );
  if args = [] then raise (Support.Argparse.Usage_error 1);
  args |> List.iter (import_feed options)
